File: reewon/Week1/방문 길이.py
def solution(dirs):
    answer = 0
    # 좌표평면을 행렬로 표현((-5,-5)를 coordinate[0][0]으로)
    # U, D, R, L 순
    # False는 해당 길이 쓰이지 않음을 표현
    coordinate = [[[False, False, False, False] for _ in range(11)] for _ in range(11)]
    i, j = 5, 5
    
    for x in dirs:
        if x == "U":
            if j+1 < len(coordinate):
                if coordinate[i][j][0] == True:
                    j += 1
                else:
                    coordinate[i][j][0] = True
                    j += 1
                    coordinate[i][j][1] = True
                    answer += 1
        elif x == "R":
            if i+1 < len(coordinate):
                if coordinate[i][j][2] == True:
                    i += 1
                else:
                    coordinate[i][j][2] = True
                    i += 1
                    coordinate[i][j][3] = True
                    answer += 1
        elif x == "D":
            if j-1 >= 0:
                if coordinate[i][j][1] == True:
                    j -= 1
                else:
                    coordinate[i][j][1] = True
                    j -= 1
                    coordinate[i][j][0] = True
                    answer += 1
        elif x == "L":
            if i-1 >= 0:
                if coordinate[i][j][3] == True:
                    i -= 1
                else:
                    coordinate[i][j][3] = True
                    i -= 1
                    coordinate[i][j][2] = True
                    answer += 1    
    return answer

File: Sjaize/Week1/방문 길이.java
// 3차원 배열을 이용하여 방문 처리를 수행할 수 있다.
class Solution {
    static int[] dy = {-1,0,0,1};
    static int[] dx = {0,1,-1,0};
    
    public int solution(String dirs) {
        boolean[][][] visited = new boolean[11][11][4];
        int y = 5, x = 5, d = 0;
        int answer = 0;
        
        for (char command : dirs.toCharArray()) {
            if (command == 'U') d = 0;
            if (command == 'R') d = 1;
            if (command == 'L') d = 2;
            if (command == 'D') d = 3;
            
            int ny = y + dy[d];
            int nx = x + dx[d];
            if (ny < 0 || ny >= 11 || nx < 0 || nx >= 11) continue;
            if (!visited[ny][nx][d]) {
                visited[ny][nx][d] = true;
                visited[y][x][3-d] = true;
                answer++;
            }
            
            y = ny;
            x = nx;
        }
        
        return answer;
    }
}
